### PR TITLE
Implementation of Frame Allocator and Subblock Allocator

### PR DESCRIPTION
**Pull request by PrimeAce123**
This pull request contains a working implementation of the system Subblock Allocator and Frame Allocator (including Dummy Allocator to setup the coremap). They have passed simple tests (which can be found in mem/mod.rs) but further testing should be done to ensure correctness of internal data structures. 

As of currently, the following features have not been implemented yet in the Subblock Allocator
1. Allocations for sizes larger than one frame will cause an AllocError (this means that the call to setup the page tables will fail because it requires a size larger than one frame)
2. We do not do frame reclamation after all the subblocks in a certain frame have been freed
3. 'deinit' for both the frame and subblock allocator have not been implemented yet

### CODE REVIEW

**Review by tonyhu-x — CHANGES_REQUESTED** [review]
Shouldn't be merged yet until we implement larger-than-one-block allocations.

File: kernel/src/mem/frame_allocator.rs [hunk 1]
```diff
@@ -1,42 +1,321 @@
...
+    placement_policy: PlacementPolicy,
+    total_number_of_frames: usize,
+}
+
+#[allow(unused)]
+pub enum PlacementPolicy {
+    Next,
+    First,
+    Best,
+}
```
[the remark this answers is not in the document]
Comment (PrimeAce123) [reply kernel/src/mem/frame_allocator.rs:94]: I got a warning from Rust about how all the enums ended with the same word; I thought that it wouldn't make too much of a difference but now that I think about it, it is definitely more clear to use the full name i.e. NextFit, FirstFit, BestFit

File: kernel/src/mem/mod.rs [hunk 2]
```diff
@@ -1,91 +1,88 @@
...
+    /// Output: The number of frames deallocated
+    fn dealloc(&mut self, ptr_to_dealloc: NonNull<u8>) -> usize;
 }
 
-struct FrameAllocatorWrapper<A: Allocator> {
-    start: NonNull<u8>,
-    frame_allocator: FrameAllocatorSolution<A>,
+struct FrameAllocatorWrapper {
+    frame_allocator: FrameAllocatorSolution,
 }
```
[the remark this answers is not in the document]
Comment (PrimeAce123) [reply kernel/src/mem/mod.rs:53]: I will definitely refactor that

File: kernel/src/mem/subblock_allocator.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                    let start_of_subblock_ptr = start_of_subblock_addr as *mut u8 as *mut ListNode;
+
+                    let next = self.list_heads[subblock_size_index].take();
+                    let new_node = ListNode { next };
+
+                    unsafe {
+                        start_of_subblock_ptr.write(new_node);
+                        self.list_heads[subblock_size_index] = Some(&mut *start_of_subblock_ptr)
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (PrimeAce123) [reply kernel/src/mem/subblock_allocator.rs:104]: Can you elaborate a bit on what you mean by pointer arithmetic? Are you referring to to just avoiding the cast to usize for the start_of_subblock_ptr part?

File: kernel/src/mem/subblock_allocator.rs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    if num_bytes > PAGE_FRAME_SIZE {
+        panic!("Requested memory size larger than page frame size");
+    }
+    for (index, size) in SUBBLOCK_SIZES.into_iter().enumerate() {
+        if num_bytes <= size {
+            return index;
+        }
+    }
+    SUBBLOCK_TYPE_COUNT
+}
```
[the remark this answers is not in the document]
Comment (PrimeAce123) [reply kernel/src/mem/subblock_allocator.rs:27]: Sounds good

File: kernel/src/mem/mod.rs [hunk 5]
```diff
@@ -220,76 +228,115 @@ impl KernelAllocator {
...
+            let num_frames_requested =
+                ((size + align).next_multiple_of(PAGE_FRAME_SIZE)) / PAGE_FRAME_SIZE;
+
+            let Ok(region) = dummy_allocator.alloc(num_frames_requested) else {
+                halt!("[KERNEL ALLOCATOR]: Unable to allocate memory according to provided layout in DummyAllocator");
+            };
+
+            // At this point, we know the allocation was successful; increment global statistics
+            let new_total_allocs = TOTAL_NUM_ALLOCATIONS.load(Ordering::Relaxed) + 1;
+            TOTAL_NUM_ALLOCATIONS.store(new_total_allocs, Ordering::Relaxed);
```
[the remark this answers is not in the document]
Comment (PrimeAce123) [reply kernel/src/mem/mod.rs:257]: I think Jack mentioned that we should try to keep allocation statistics (such as how many allocations were made, total size of allocations, etc). I think later we will need these metrics, but for now, I think I can probably replace this with a bool like you suggested; we can add the metrics in a later pull request.

File: kernel/src/mem/subblock_allocator.rs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                assert!(size_of::<ListNode>() <= SUBBLOCK_SIZES[subblock_size_index]);
+
+                if !self.frame_allocator.has_room(1) {
+                    return Err(AllocError);
+                }
+
+                let new_frame = match self.frame_allocator.alloc(1) {
+                    Err(AllocError) => return Err(AllocError),
+                    Ok(v) => v,
+                };
```
[the remark this answers is not in the document]
Comment (PrimeAce123) [reply kernel/src/mem/subblock_allocator.rs:87]: Sounds good

File: kernel/src/mem/subblock_allocator.rs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        }
+    }
+
+    /// Allocate a fixed size block for the size requested by 'layout'
+    ///
+    pub fn allocate(&mut self, layout: Layout) -> Result<*mut u8, AllocError> {
+        let subblock_size_index = get_best_subblock_size_idx(layout.size());
+
+        if subblock_size_index == SUBBLOCK_TYPE_COUNT {
+            println!(
```
[the remark this answers is not in the document]
Comment (PrimeAce123) [reply kernel/src/mem/subblock_allocator.rs:54]: Sounds good

**Comment by PrimeAce123** [comment]
> Shouldn't be merged yet until we implement larger-than-one-block allocations.

Implemented this feature in the latest commit

In terms of testing, I did some playing around and I'm pretty sure the current tests in "mem/mod.rs" don't use the allocator we have written; it uses some other global allocator (I think its the default Rust one but I'm not sure). This is something we should address with the other groups to try and make sure all their tests use our allocator (as a temporarily solution, I've been running the tests in the main function and seeing if they pass)

**Comment by tonyhu-x** [comment]
> > Shouldn't be merged yet until we implement larger-than-one-block allocations.
> 
> Implemented this feature in the latest commit
> 
> In terms of testing, I did some playing around and I'm pretty sure the current tests in "mem/mod.rs" don't use the allocator we have written; it uses some other global allocator (I think its the default Rust one but I'm not sure). This is something we should address with the other groups to try and make sure all their tests use our allocator (as a temporarily solution, I've been running the tests in the main function and seeing if they pass)

The tests don't run our allocator, and they wouldn't be able to. The tests themselves run on the host, i.e. whichever OS we're writing our code in. Therefore, we can't directly initialize the `KernelAllocator` the same way we do when the code runs on QEMU. Even if we could, since the subblock allocator itself is not a global allocator, in our unit tests we should call its functions directly. Actually doing it is a bit difficult though, since the subblock allocator depends on `FrameAllocator`, so we can either initialize a `FrameAllocator`, giving it a pre-allocated pool of memory, or pass a fake `FrameAllocator` to our subblock allocator.

File: kernel/src/mem/mod.rs [hunk 8]
```diff
@@ -220,76 +197,111 @@ impl KernelAllocator {
...
-            subblock_allocator.deallocate(NonNull::new_unchecked(ptr), layout);
+        TOTAL_NUM_DEALLOCATIONS.fetch_add(1, Ordering::Relaxed);
+    }
+}
 
-            if !subblock_allocator.is_empty() {
-                return;
-            }
+#[cfg(test)]
+mod test {
```
Comment (tonyhu-x) [inline kernel/src/mem/mod.rs:268]: `mod tests` may be more conventional.

**Comment by tonyhu-x** [comment]
Before we merge the code, we should also remove the `println` statements.

File: kernel/src/mem/frame_allocator.rs [hunk 9]
```diff
@@ -1,42 +1,96 @@
...
+            NonNull::new(unsafe { self.start.as_ptr().add(range.start * PAGE_FRAME_SIZE) })
+                .ok_or(AllocError)?,
+            range.len() * PAGE_FRAME_SIZE,
+        ))
     }
 
-    fn alloc(&mut self, frames: usize) -> Option<Range<usize>> {
-        if self.first_unused_frame + frames >= self.max_frames {
-            return None;
+    fn dealloc(&mut self, ptr_to_dealloc: NonNull<u8>) -> usize {
```
[the remark this answers is not in the document]
Comment (PrimeAce123) [reply kernel/src/mem/frame_allocator.rs:69]: Sounds good

File: kernel/src/mem/mod.rs (hunk 8 above)
Comment (PrimeAce123) [reply kernel/src/mem/mod.rs:268]: Sounds good

File: kernel/src/mem/frame_allocator.rs [hunk 10]
```diff
@@ -1,42 +1,145 @@
...
-        Self {
-            alloc,
-            max_frames,
-            first_unused_frame: 0,
+impl FrameAllocator for FrameAllocatorSolution {
+    fn new_in(start: NonNull<u8>, core_map: Box<[CoreMapEntry]>) -> Self {
+        FrameAllocatorSolution {
+            start,
+            core_map,
+            placement_algorithm: next_fit,
```
Comment (tonyhu-x) [inline kernel/src/mem/frame_allocator.rs:41]: I just realized that there is no way for the `KernelAllocator` to provide the placement algorithm. Instead, the choice is fixed here. We should change it. Further, I don't see a reason why the `FrameAllocator` trait has to define a constructor function. It just needs the signatures for `alloc` and `dealloc`. The constructor should be defined in its own `impl` block.

File: kernel/src/mem/frame_allocator.rs [hunk 11]
```diff
@@ -1,42 +1,145 @@
...
-    fn new_in(alloc: A, max_frames: usize) -> Self
-    where
-        Self: Sized,
-    {
-        Self {
-            alloc,
-            max_frames,
-            first_unused_frame: 0,
+impl FrameAllocator for FrameAllocatorSolution {
+    fn new_in(start: NonNull<u8>, core_map: Box<[CoreMapEntry]>) -> Self {
```
Comment (tonyhu-x) [inline kernel/src/mem/frame_allocator.rs:37]: I think we should rename the constructor to `new`. It seems that the standard library uses `new_in` as a conventional name for constructors that take a custom allocator. This is not what happens here.

File: kernel/src/mem/frame_allocator.rs (hunk 10 above)
Comment (PrimeAce123) [reply kernel/src/mem/frame_allocator.rs:41]: Got it; that does make a lot more sense

File: kernel/src/mem/frame_allocator.rs (hunk 11 above)
Comment (PrimeAce123) [reply kernel/src/mem/frame_allocator.rs:37]: Sounds good

**Comment by tonyhu-x** [comment]
Refactored the code a bit more (read the commit body for details). Let me know what you think.

**Review by tonyhu-x — APPROVED** [review]

**Comment by PrimeAce123** [comment]
Before merging, CoreMapEntry in frame_allocator.rs needs to be refactored to not use bitbybit but instead bitfield (since bitbybit dependency is removed)

**Pull request merged**